docs/specification.md

```diff
@@ -57,33 +51,83 @@ Compile-time value parameters are captured by the `Config` type (parameterized b
 data Config proposalMetadata = Config
   { cDaoName :: Text
   -- ^ Name of the DAO.
-  , cDaoDescription :: Text
+  , cDaoDescription :: Markdown
   -- ^ Description of the DAO.
-  , cProposalCheck :: Lambda (Natural, Proposal proposalMetadata) Bool
+  , cUnfrozenTokenMetadata :: FA2.TokenMetadata
+  -- ^ FA2 metadata for unfrozen token.
+  , cFrozenTokenMetadata :: FA2.TokenMetadata
+  -- ^ FA2 metadata for frozen token.
+  , cProposalCheck :: forall s
+      . (ProposeParams proposalMetadata) & s
+    :-> Bool & s
   -- ^ A lambda used to verify whether a proposal can be submitted.
   -- It checks 2 things: the proposal itself and the amount of tokens frozen upon submission.
   -- It allows the DAO to reject a proposal by arbitrary logic and captures bond requirements
-  -- (i. e. how much one should "freeze" in order to propose).
-  , cUnfrozenTokenMetadata :: FA2Metadata
-  -- ^ FA2 metadata for unfrozen token.
-  , cFrozenTokenMetadata :: FA2Metadata
-  -- ^ FA2 metadata for frozen token.
-  , cDecisionLambda :: Lambda
-      (Proposal proposalMetadata, Storage proposalMetadata)
-      (List Operation, Storage proposalMetadata)
-  -- ^ The decision lambda is executed based on a successful proposal.
-  , cRejectedProposalReturnValue :: Lambda Natural Natural
+  , cRejectedProposalReturnValue :: forall s
+      . (Proposal proposalMetadata) & s
+    :-> ("slash_amount" :! Natural) & s
   -- ^ When a proposal is rejected, the value that voters get back can be slashed.
   -- This lambda specifies how many tokens will be received.
   -- For example, if Alice freezes 100 tokens to vote and this lambda divides the value by 2,
   -- only 50 tokens will be unfrozen and other 50 tokens will be burnt.
+  , cDecisionLambda :: forall s
+      . (Proposal proposalMetadata) & (Storage proposalMetadata) & s
+    :-> (List Operation, Storage proposalMetadata) & s
+  -- ^ The decision lambda is executed based on a successful proposal.
+  , cMaxProposals :: Natural
+  , cMaxVotes :: Natural
+  , cMaxQuorumThreshold :: Natural
+  , cMinQuorumThreshold :: Natural
+  , cMaxVotingPeriod :: Natural
+  , cMinVotingPeriod :: Natural
   }
 ```
 
 The `FA2Metadata` type matches `token_metadata` defined in FA2.
-The `Proposal` type is defined below.
-`Storage` is the full storage type of the contract.
 
+`Storage` is the full storage type of the contract:
+
+```haskell
+-- | Voting period in seconds
+type VotingPeriod = Natural
+
+-- | Quorum threshold of a proposal that needs to be met to be accepted
+-- quorum_threshold = upvotes - downvotes
+type QuorumThreshold = Natural
+
+data Storage = Storage
```

Comment:
I'd rather not specify the exact storage type here because:
1. If we specify it here, we will have to maintain it up to date with code and most likely we'll forget to do it at some point. Since not everything is implemented yet, the storage type will likely change many times.
2. This specification is supposed to explain what the contract does and how one can use it. "Usage" basically boils down to calling entrypoints (that's why we specify their exact types) and constructing the initial storage. A large part of the initial storage is predefined. For example, `sProposalIdCounter` is always 0 initially. There are some values that should be provided explicitly, we already list them in "Runtime specification". One may also inspect storage fields directly, that's why we say that "The storage of the contract must have annotations for all fields and must be documented to make its interpretation easy for users".
3. As I wrote in another comment, this specification shouldn't be deeply technical. The fact that we, for example, store proposal IDs sorted by date is really just an implementation detail. Keep in mind that we also have auto-generated documentation that contains the whole storage type which is updated automatically. Those who want to interact with the contract at lower level (e. g. implement their own decision lambda) can refer to autogenerated documentation instead.

Reply:
I agree that it can be a pain to maintain. Initially, I thought this may be useful as an example of the implementation of `Storage`. But it can be confusing and too technical.

I've removed this for now, but I still keep the type `Proposal` since it is used in lambda functions specified in `Config` and would be useful to understand how those functions work.

Reply:
`Proposal` is ok, it's less likely to change while `Storage` gathers data from all components of the contract, so it is expected to change sometimes.